fiaas_deploy_daemon/deployer/kubernetes/role_binding.py

```diff
@@ -0,0 +1,45 @@
+import logging
+
+from k8s.models.common import ObjectMeta
+from k8s.models.role_binding import RoleBinding, RoleRef, Subject
+from k8s.client import NotFound
+from fiaas_deploy_daemon.specs.models import AppSpec
+from fiaas_deploy_daemon.deployer.kubernetes.owner_references import OwnerReferences
+
+LOG = logging.getLogger(__name__)
+
+
+class RoleBindingDeployer:
+    def __init__(self, config, owner_references):
+        self._config = config
+        self._owner_references: OwnerReferences = owner_references
+
+    def deploy(self, app_spec: AppSpec):
+        self._create_role_bindings(app_spec, self._config.list_of_roles, "Role", 1)
+        self._create_role_bindings(app_spec, self._config.list_of_cluster_roles, "ClusterRole", 1 + len(self._config.list_of_roles))
+
+    def _create_role_bindings(self, app_spec: AppSpec, roles_list, role_kind, counter):
+        LOG.info("Creating RoleBindings for %s", app_spec.name)
+        namespace = app_spec.namespace
+
+        service_account_name = app_spec.name
+
+        for role_name in roles_list:
+            role_binding_name = f"{app_spec.name}-{counter}"
+            try:
+                role_binding = RoleBinding.get(role_binding_name, namespace)
+            except NotFound:
+                role_binding = RoleBinding()
+
+            role_binding.metadata = ObjectMeta(
+                name=role_binding_name, namespace=namespace
+            )
+
+            role_ref = RoleRef(kind=role_kind, apiGroup="rbac.authorization.k8s.io", name=role_name)
+            subject = Subject(kind="ServiceAccount", name=service_account_name, namespace=namespace)
+            role_binding.roleRef = role_ref
+            role_binding.subjects = [subject]
+            self._owner_references.apply(role_binding, app_spec)
+            role_binding.save()
+
+            counter += 1
```

Comment:
Two points about the deployer logic here:

1) If a existing role in `list-of-roles` or `list-of-cluster-roles` is removed, the role bindings binding that role to deployed applications service accounts are not removed. I think in that case the (cluster)rolebindings probably should be deleted, to avoid applications retaining privileges they should not have.

2) Because [roleRef is immutable](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding), the rolebinding naming scheme used here will cause problems when there are changes to the order of the configured roles or clusterroles. Let's say fiaas-deploy-daemon runs with `list-of-roles: [foo, bar]`. Deployed applications have bindings `$app-1` for role `foo` and `$app-2` for role `bar`. If `foo` is removed from the configuration, on the next deploy fiaas-deploy-daemon will try to update rolebinding `$app-1` to bind role `bar`, and the deploy will fail because the API will reject changing the roleRef for that resource. 

Maybe one way to handle these cases may be e.g. to have the rolebinding deployer list all rolebindings and clusterrolebindings with selector `app=$appname,fiaas/deployed_by` (see comment about labels above), compare the desired roles and clusterroles with the existing roles, update the bindings which bind desired roles/clusterroles and delete bindings for roles which are no longer desired. Maybe also do a [check similar to for serviceaccounts](https://github.com/fiaas/fiaas-deploy-daemon/blob/master/fiaas_deploy_daemon/deployer/kubernetes/service_account.py#L78-L82) to avoid accidentally deleting bindings managed by other systems. There are probably other ways to do it as well, this was just my first thought.

Reply:
Hey @oyvindio, I've just changed the implementation. Can you check? 😄 

I want to discuss first if we are happy with this new implementation before starting to work on the tests 😄 

Right now, I'm getting the list of role_bindings with the label `app: app_name` and then I delete those `owned_by_fiaas` that do not have a role or cluster role provided.

Then, I iterate over the roles and role_bindings and, for any of them, first, I check in the list if exists any role_binding with the expected role or cluster role. I just update it if it exists and it's `owned_by_fiaas`. If it exists and it's not owned_by_fiaas, I log a message.

If it does not exist, I search for an `app-name-{i}` that does not exist and just create it 😄